import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CFindService.java

```diff
@@ -0,0 +1,153 @@
+import org.dcm4che3.net.Status;
+import org.dcm4che3.net.pdu.PresentationContext;
+import org.dcm4che3.net.service.BasicCFindSCP;
+import org.dcm4che3.net.service.DicomServiceException;
+import org.dcm4che3.util.TagUtils;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class CFindService extends BasicCFindSCP {
+
+  private static Logger log = LoggerFactory.getLogger(CFindService.class);
+
+  private IDicomWebClient dicomWebClient;
+
+  CFindService(IDicomWebClient dicomWebClient) {
+    super(UID.StudyRootQueryRetrieveInformationModelFIND);
+    this.dicomWebClient = dicomWebClient;
+  }
+
+  private static HashMap<String, JSONObject> uniqueResults(List<JSONArray> responses) {
+    HashMap<String, JSONObject> uniqueResults = new HashMap<>();
+    for (JSONArray response : responses) {
+      for (Object result : response) {
+        JSONObject resultJson = (JSONObject) result;
+        String key = getResultKey(resultJson);
+        if (!uniqueResults.containsKey(key)) {
+          uniqueResults.put(key, resultJson);
+        }
+      }
+    }
+    return uniqueResults;
+  }
+
+  private static String getResultKey(JSONObject jsonObject) {
+    return AttributesUtil.getTagValueOrNull(jsonObject,
+        TagUtils.toHexString(Tag.StudyInstanceUID)) + "_" +
+        AttributesUtil.getTagValueOrNull(jsonObject,
+            TagUtils.toHexString(Tag.SeriesInstanceUID)) + "_" +
+        AttributesUtil.getTagValueOrNull(jsonObject,
+            TagUtils.toHexString(Tag.SOPInstanceUID));
+  }
+
+  @Override
+  public void onDimseRQ(Association association,
+      PresentationContext presentationContext,
+      Dimse dimse,
+      Attributes request,
+      Attributes keys) throws IOException {
+    if (dimse != Dimse.C_FIND_RQ) {
+      throw new DicomServiceException(Status.UnrecognizedOperation);
+    }
+
+    MonitoringService.addEvent(Event.CFIND_REQUEST);
+
+    CFindTask task = new CFindTask(association, presentationContext, request, keys);
+    association.getApplicationEntity().getDevice().execute(task);
+  }
+
+  private class CFindTask extends DimseTask {
+
+    private final Attributes keys;
+
+    private CFindTask(Association as, PresentationContext pc,
+        Attributes cmd, Attributes keys) {
+      super(as, pc, cmd);
+
+      this.keys = keys;
+    }
+
+    @Override
+    public void run() {
+      try {
+        if (canceled) {
+          throw new CancellationException();
+        }
+        runThread = Thread.currentThread();
+
+        String[] qidoPaths = AttributesUtil.attributesToQidoPathArray(keys);
+        List<JSONArray> qidoResults = new ArrayList<>();
+        for (String qidoPath : qidoPaths) {
+          if (canceled) {
+            throw new CancellationException();
+          }
+          log.info("CFind QidoPath: " + qidoPath);
+          try {
+            MonitoringService.addEvent(Event.CFIND_SUB_QIDORS_REQUEST);
```

Comment:
I'm a bit confused with the name "CFIND_SUB_QIDORS_REQUEST", not sure what SUB means here, can we shorten these to "CFIND_QIDORS_REQUEST", etc?

Reply:
Ok, removed.